### PR TITLE
[SPARK-31387] Handle unknown operation/session ID in HiveThriftServer2Listener

### PR DESCRIPTION
**Pull request by alismess-db**
### What changes were proposed in this pull request?

The update methods in HiveThriftServer2Listener now check if the parameter operation/session ID actually exist in the `sessionList` and `executionList` respectively. This prevents NullPointerExceptions if the operation or session ID is unknown. Instead, a warning is written to the log.

Also, in HiveSessionImpl.close(), we catch any exception thrown by `operationManager.closeOperation`. If for any reason this throws an exception, other operations are not prevented from being closed.

### Why are the changes needed?

The listener's update methods would throw an exception if the operation or session ID is unknown. In Spark 2, where the listener is called directly, this hampers with the caller's control flow. In Spark 3, the exception is caught by the ListenerBus but results in an uninformative NullPointerException.

In HiveSessionImpl.close(), if an exception is thrown when closing an operation, all following operations are not closed.

### Does this PR introduce any user-facing change?

No

### How was this patch tested?

Unit tests

### CODE REVIEW

**Comment by juliuszsompolski** [comment]
cc @wangyum 
> In Spark 2, where the listener is called directly, this hampers with the caller's control flow.
This exception could happen if an operation in ERROR or CANCELED state that was never closed by the client was already evicted from the listener state (as not-active) when the session was later closed.

**Comment by juliuszsompolski** [comment]
It is more serious in Spark 2, where the exception from the listener would bubble up all the way to session, but IMO we could also catch it in Spark 3 listener, as this PR proposes.

File: sql/hive-thriftserver/src/test/scala/org/apache/spark/sql/hive/service/cli/session/HiveSessionImplSuite.scala [hunk 1]
```diff
@@ -0,0 +1,68 @@
...
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package org.apache.spark.sql.hive.service.cli.session
+
+import org.apache.hadoop.hive.conf.HiveConf
+import org.apache.hive.service.cli.operation.{GetCatalogsOperation, OperationManager}
+import org.apache.hive.service.cli.session.{HiveSessionImpl, SessionManager}
+import org.apache.hive.service.rpc.thrift.TProtocolVersion
```
Comment (juliuszsompolski) [inline sql/hive-thriftserver/src/test/scala/org/apache/spark/sql/hive/service/cli/session/HiveSessionImplSuite.scala:22]: https://github.com/apache/spark/pull/28155/checks?check_run_id=571648817
```
[ERROR] [Error] /home/runner/work/spark/spark/sql/hive-thriftserver/src/test/scala/org/apache/spark/sql/hive/service/cli/session/HiveSessionImplSuite.scala:22: object rpc is not a member of package org.apache.hive.service
[ERROR] [Error] /home/runner/work/spark/spark/sql/hive-thriftserver/src/test/scala/org/apache/spark/sql/hive/service/cli/session/HiveSessionImplSuite.scala:37: not found: value TProtocolVersion
[ERROR] two errors found
[ERROR] Failed to execute goal net.alchim31.maven:scala-maven-plugin:4.3.0:testCompile (scala-test-compile-first) on project spark-hive-thriftserver_2.12: Execution scala-test-compile-first of goal net.alchim31.maven:scala-maven-plugin:4.3.0:testCompile failed.: CompileFailed -> [Help 1]
```
You can use `ThriftserverShimUtils.testedProtocolVersions(0)` to make it work with both hive-1.2 and hive-2.3 profile.
You must import TProtocolVersion from

**Comment by juliuszsompolski** [comment]
@wangyum @gatorsmile could you take a look?
It improves the robustness of Thriftserver session / Thriftserver listener against uncaught exceptions.
In Spark 2.4, when the listener was running in the same thread as execution, this could cause some sessions to not be properly cleaned up if an exception was thrown from one of the operations in the session during session close.

**Review by juliuszsompolski — APPROVED** [review]
LGTM

**Comment by gatorsmile** [comment]
ok to test

**Comment by gatorsmile** [comment]
LGTM

Thanks! Merged to master/3.0

**Comment by dongjoon-hyun** [comment]
Hi, @alismess-db , @juliuszsompolski , @gatorsmile .

Is this tested with `Hive 1.2` profile?

According to the Jenkins result, this seems to break `Hive 1.2` profile. Could you take a look at this?
- https://amplab.cs.berkeley.edu/jenkins/view/Spark%20QA%20Test%20(Dashboard)/job/spark-branch-3.0-test-maven-hadoop-2.7-hive-1.2/324/testReport/org.apache.spark.sql.hive.thriftserver/SparkSQLEnvSuite/SPARK_29604_external_listeners_should_be_initialized_with_Spark_classloader/history/

Screen Shot 2020-05-14 at 11 09 25 AM

**Comment by dongjoon-hyun** [comment]
Ur, I found that this breaks `hive-2.3`, too. In short, this seems to break all Maven tests in `branch-3.0`.

Screen Shot 2020-05-14 at 11 49 34 AM

**Comment by dongjoon-hyun** [comment]
Ur, it turns out that `master` branch Maven jobs are broken, too.
Screen Shot 2020-05-14 at 11 59 43 AM


I'll revert this patch from master/3.0. Please make another PR and test with Maven.

**Comment by dongjoon-hyun** [comment]
FYI, two JDK11 jobs are recovered. I'm looking at the other failures like https://github.com/apache/spark/pull/28532 .
- https://amplab.cs.berkeley.edu/jenkins/view/Spark%20QA%20Test%20(Dashboard)/job/spark-branch-3.0-test-maven-hadoop-2.7-hive-2.3-jdk-11/
- https://amplab.cs.berkeley.edu/jenkins/view/Spark%20QA%20Test%20(Dashboard)/job/spark-branch-3.0-test-maven-hadoop-3.2-hive-2.3-jdk-11/

**Pull request closed without merging**